### ISSUE TITLE
Data source not updated when it depends on a different resource

### ISSUE THREAD

**Issue by sblask** [hashicorp/terraform#27589]
I already reported my problem here: https://github.com/hashicorp/terraform-provider-aws/issues/17176 but got pointed to this repository instead.

### Terraform Version

```
Terraform v0.13.5

provider registry.terraform.io/hashicorp/aws v3.21.0
provider registry.terraform.io/hashicorp/random v3.0.0
provider registry.terraform.io/hashicorp/template v2.2.0
provider registry.terraform.io/hashicorp/vault v2.16.0
```

### Terraform Configuration Files

```terraform
data "aws_iam_group" "developers" {
  group_name = "developer"
}

locals {
  developers = {
    for user in data.aws_iam_group.developers.users
    :
    replace(split("@", user.user_name)[0], ".", "-") => { arn = user.arn, name = user.user_name }
  }
}

resource "aws_iam_user_policy" "assume" {
  for_each = local.developers

  name = "${each.key}"
  user = each.value.name

  policy = data.aws_iam_policy_document.assume[each.key].json
}

data "aws_iam_policy_document" "assume" {
  for_each = local.developers

  statement {
    sid       = "AssumeReplRole"
    actions   = ["sts:AssumeRole"]
    resources = [aws_iam_role.role[each.key].arn] # I run into the same problem when I remove this but have an explicit depends_on to aws_iam_role.role
  }
}

resource "aws_iam_role" "role" {
  for_each = local.developers
```

### Expected Behavior

I should always be able to plan/apply this

### Actual Behavior

When the aws_iam_group contains more users than during a previous run, like I have 12 users, apply and add new.user@example.com I get:

```
Error: Invalid index

  on xyz line 67, in resource "aws_iam_user_policy" "assume":
  67:   policy = data.aws_iam_policy_document.assume[each.key].json
    |----------------
    | data.aws_iam_policy_document.assume is object with 12 attributes
    | each.key is "new-user"
```

No matter how many users there are, it always shows the previous number

### Steps to Reproduce

1. terraform apply
1. add user
1. terraform plan

### References

https://github.com/hashicorp/terraform-provider-aws/issues/17176

**Comment by jbardin** [comment]
Hi @sblask 

Thanks for filing the issue. Can you reproduce this with the current terraform release? (0.14.5 is the latest)
This is likely a problem that was not solve-able with the separate "refresh" phase of terraform operation, but that has been removed in 0.14, with resources being read on-demand during plan.

Thanks!

**Comment by sblask** [comment]
Seems like this has been fixed in version 0.14! Thank you.
